File: All Algorithms/Sorting Algorithm in Python/bubbleSort.py
# BUBBLE SORT
# Compare two elements and sort them.. compare next two and sort them.. and so on

# For Example: 2 5 1 6 3 7        n = 6
# Step 1: Compare 2 & 5.. they are already sorted.. so continue.. ResultArray: 2 5 1 6 3 7
# Step 2: Compare 5 & 1.. they are not sorted.. so swap.. ResultArray: 2 1 5 6 3 7
# Step 3: Compare 5 & 6.. they are already sorted.. so continue.. ResultArray: 2 1 5 6 3 7
# Step 4: Compare 6 & 3.. they are not sorted.. so swap.. ResultArray: 2 1 5 3 6 7
# Step 5: Compare 6 & 7.. they are already sorted.. so continue.. ResultArray: 2 1 5 3 6 7

# After this many steps largest element has been placed at correct position
# Here we had to look for n elements.. for sorting 1 element
# Do this process again & again till the array is completely sorted

# So now one element is sorted now we have to look for (n-1) elements.. next (n-2) then (n-3).. until 1
# So the time complexity: n+(n-1)+(n-2)+...+2+1 = n*(n-1)/2 = n^2

# Time Complexity: O(n^2)

# ------------------------------------------------- CODE -------------------------------------------------
import random

def bubbleSort(sortingList):

    for i in range(len(sortingList)):
        for j in range(len(sortingList) - i - 1):
            if sortingList[j] > sortingList[j+1]:
                sortingList[j], sortingList[j+1] = sortingList[j+1], sortingList[j]

    return sortingList

sortList = random.sample(range(1, 11111), 9999)
print(bubbleSort(sortList))

File: Basic Programming/BubbleSort.cpp
/*
  Bubble Sort Algorithm in C++

  Time Complexity : O(n^2)
  Space Complexity : O(1)

  (Sorting in increasing order)
  -> Swapping the two elements if right element is greater than left element (repeat until gets a sorted array i.e. n-1 times).
  -> After each traversal we get one element in its right most correct position.
*/

#include<iostream>
using namespace std;

// function to swap two elements
void swap(int *x, int *y) {
  int temp = *x;
  *x = *y;
  *y = temp;
}

// function to print array
void printArr(int arr[], int n) {
  for(int i=0; i<n; i++) {
    cout<<arr[i]<<" ";
  }
  cout<<endl;
}

// function for Bubble Sort
int *bubbleSort(int arr[], int n) {
  for(int i=1; i<n; i++) {
    for(int j=0; j<n-i; j++) {
      if(arr[j] > arr[j+1]) {
        swap(arr[j], arr[j+1]);
      }
    }
  }
  return arr;
}

int main() {
  int unsortedArr[7] = {3, 7, 2, 8, 1, 0, 9};
  int size = sizeof(unsortedArr)/sizeof(unsortedArr[0]);

  cout<<"Unsorted Array : ";
  printArr(unsortedArr, size);

  int *sortedArr = bubbleSort(unsortedArr, size);
  cout<<"Sorted Array : ";
  printArr(sortedArr, size);

  return 0;
}
